src/energy_distribution.F90

```diff
@@ -0,0 +1,388 @@
+
+    integer :: i, k, l
+    integer :: n_energy_in
+    integer :: n_energy_out
+    real(8) :: r              ! interpolation factor on incoming energy
+    real(8) :: E_i_1, E_i_K   ! endpoints on outgoing grid i
+    real(8) :: E_i1_1, E_i1_K ! endpoints on outgoing grid i+1
+    real(8) :: E_1, E_K       ! endpoints interpolated between i and i+1
+    real(8) :: E_l_k, E_l_k1  ! adjacent E on outgoing grid l
+
+    ! Determine number of incoming/outgoing energies
+    n_energy_in = size(this%energy_in)
+    n_energy_out = size(this%energy_out, 1)
+
+    ! determine index on incoming energy grid and interpolation factor
+    i = binary_search(this%energy_in, size(this%energy_in), E_in)
+    r = (E_in - this%energy_in(i)) / &
+         (this%energy_in(i+1) - this%energy_in(i))
+
+    ! Sample outgoing energy bin
+    k = 1 + int(n_energy_out * prn())
+
+    ! Determine E_1 and E_K
+    E_i_1 = this%energy_out(1, i)
+    E_i_K = this%energy_out(n_energy_out, i)
+
+    E_i1_1 = this%energy_out(1, i+1)
+    E_i1_K = this%energy_out(n_energy_out, i+1)
+
+    E_1 = E_i_1 + r*(E_i1_1 - E_i_1)
+    E_K = E_i_K + r*(E_i1_K - E_i_K)
+
+    ! Randomly select between the outgoing table for incoming energy E_i and
+    ! E_(i+1)
+    if (prn() < r) then
+      l = i + 1
+    else
+      l = i
+    end if
+
+    ! Determine E_l_k and E_l_k+1
+    E_l_k  = this%energy_out(k, l)
+    E_l_k1 = this%energy_out(k+1, l)
+
+    ! Determine E' (denoted here as E_out)
+    E_out  = E_l_k + prn()*(E_l_k1 - E_l_k)
+
+    ! Now interpolate between incident energy bins i and i + 1
+    if (l == i) then
+      E_out = E_1 + (E_out - E_i_1)*(E_K - E_1)/(E_i_K - E_i_1)
+    else
+      E_out = E_1 + (E_out - E_i1_1)*(E_K - E_1)/(E_i1_K - E_i1_1)
+    end if
+  end function equiprobable_sample
+
+  function level_inelastic_sample(this, E_in) result(E_out)
+    class(LevelInelastic), intent(in) :: this
+    real(8), intent(in) :: E_in
+    real(8) :: E_out
+
+    E_out = this%mass_ratio*(E_in - this%threshold)
+  end function level_inelastic_sample
+
+  function continuous_sample(this, E_in) result(E_out)
+    class(ContinuousTabular), intent(in) :: this
+    real(8), intent(in) :: E_in
+    real(8) :: E_out
+
+    integer :: i, k, l
+    integer :: n_energy_in
+    integer :: n_energy_out
+    real(8) :: r              ! interpolation factor on incoming energy
+    real(8) :: r1             ! random number on [0,1)
+    real(8) :: frac           ! interpolation factor on outgoing energy
+    real(8) :: E_i_1, E_i_K   ! endpoints on outgoing grid i
+    real(8) :: E_i1_1, E_i1_K ! endpoints on outgoing grid i+1
+    real(8) :: E_1, E_K       ! endpoints interpolated between i and i+1
+    real(8) :: E_l_k, E_l_k1  ! adjacent E on outgoing grid l
+    real(8) :: p_l_k, p_l_k1  ! adjacent p on outgoing grid l
+    real(8) :: c_k, c_k1      ! cumulative probability
+    logical :: histogram_interp
+
+    ! read number of interpolation regions and incoming energies
```

Comment:
It seems to be our prevailing style to capitalize comments.  Please capitalize this one, and there are a few others that need it in this file.